lib/iris/fileformats/netcdf.py

```diff
@@ -1593,57 +1725,158 @@ def _get_dim_names(self, cube):
 
         Returns:
-            List of dimension names with length equal the number of dimensions
-            in the cube.
+            cube_dimensions, mesh_dimensions
+            * cube_dimensions (list of string):
+                A lists of dimension names for each dimension of the cube
+            * mesh_dimensions (list of string):
+                A list of the mesh dimensions of the attached mesh, if any.
+
+        ..note::
+            One of the mesh dimensions will typically also appear in the cube
+            dimensions.
 
         """
         dimension_names = []
-        for dim in range(cube.ndim):
-            coords = cube.coords(dimensions=dim, dim_coords=True)
-            if coords:
-                coord = coords[0]
 
-                dim_name = self._get_coord_variable_name(cube, coord)
-                # Add only dimensions that have not already been added.
-                if coord not in self._dim_coords:
-                    # Determine unique dimension name
-                    while (
-                        dim_name in self._existing_dim
-                        or dim_name in self._name_coord_map.names
-                    ):
-                        dim_name = self._increment_name(dim_name)
+        def record_dimension(dim_name, length, matching_coords=[]):
+            """
+            Record a file dimension, its length and associated coordinates.
 
-                    # Update names added, current cube dim names used and
-                    # unique coordinates added.
-                    self._existing_dim[dim_name] = coord.shape[0]
-                    dimension_names.append(dim_name)
-                    self._dim_coords.append(coord)
+            If the dimension has been seen already, check that it's length
+            matches the earlier finding.
+
+            """
+            if dim_name not in self._existing_dim:
+                self._existing_dim[dim_name] = length
+            else:
+                # Just make sure we never re-write one.
+                # TODO: possibly merits a proper Exception with message
+                assert self._existing_dim[dim_name] == length
+
+            # Add new coords (mesh or dim) to the already-seen list.
+            for coord in matching_coords:
+                if coord not in self._dim_names_and_coords.coords:
+                    self._dim_names_and_coords.append(dim_name, coord)
+
+            # Add the latest name to the returned list
+            dimension_names.append(dim_name)
+
+        # Get info on mesh, first.
+        dimension_names.clear()
+        mesh = cube.mesh
+        if mesh is None:
+            cube_mesh_dim = None
+        else:
+            # Identify all the mesh dimensions.
+            # NOTE: one of these will be a cube dimension, but that one does not
+            # get any special handling.  We *do* want to list/create them in a
+            # definite order (face,edge,node), and before non-mesh dimensions.
```

Comment:
@pp-mo Here you document that the `definite order` is `face,edge,node`, and yet `MESH_LOCATIONS = ("node", "edge", "face")`

Reply:
:+1: oops